docs/source/native/functions.rst

```diff
@@ -5,33 +5,148 @@ The vast majority of GNUT modding within northstar will be done through function
+
+If a function is lacking a ``return`` statement or a return value, it will return ``null``. For example:
 
 .. code-block:: javascript
 
   void function ThisDoesStuff()
+  {
+    switch( RandomInt( 3 ) )
+    {
+      case 0:
+        return
+      case 1:
+        return null
+      case 2:
+        break
+    }
+    // only if a 2 was rolled, code here will be executed.
+    // because the function ends without a return statement, null is getting returned implicitly 
+  }
+
+Optional Parameters
+-----
+
+Optional parameters aren't required to call the function and will be assigned a default value if nothing was passed.
+
+To make a parameter optional, add a default after the parameter
+
+.. code-block:: javascript
+
+  void function OptionalExample( string msg = "default parameter", optional2 = 1 )
+  {
+    printt( msg )
+  }
+
+  OptionalExample( "passed parameter" ) // passed parameter
+  OptionalExample() // default parameter
+
+Optional parameters must be declared after all required parameters.
+
+Passing Functions as parameters
+-----
+
+If you want to pass a function as a parameter to another function, for example as a callback set their type as ``functionref( [parameters] )``.
+
+.. code-block:: javascript
 
+  void function FnLiteral( int req, int opt = 2 )
+  {
+    print( req + opt )
+  }
+
+  void function CallLiteral( functionref( int, int ) literal )
+  {
+    literal( RandomInt( 5 ) )
+  }
+
+  CallLiteral( FnLiteral )
+
+Calling Functions
+-----
+
+You can call functions with opening and closing brackets containing all parameters or with the call function.
+
+You can also call functions with an array of parameters
+
+.. code-block:: javascript
+
+  void function FnLiteral( int opt = 2, int opt2 = 2 )
+  {
+    print( opt + opt2 )
+  }
+
+  FnLiteral() // 4
+  FnLiteral( 1, 2 ) // 3
+  FnLiteral.call( 3, 4 ) // 7
+
+  array<int> args = [ 6, 7 ]
+  FnLiteral.acall( args ) // 13
+
+Implicit parameters
+-----
+
+If you don't know how many parameters you get at compile time, you can use implicit parameters.
+
+.. code-block:: javascript
+
+  void function XParameters( string required, string optional = "", ... )
+  {
+    for( int i; i < vargc, i++)
+    {
+      var parameter = vargv[i]
+      print( parameter )
+    }
+  }
+
+  XParameters( "req", "optional", 1, 2, [ 3, 4, 5 ], { tableKey = "string" } ) // prints 1, 2, [array instance], [table instance]
+  XParameters( "req", "opt" )
+
+Locals relocalisation
+-----
+
+It's not possible to use locals from a parent function, but you can "relocate" locals to an anonymous function
+
+.. code-block:: javascript
+
+  void function ParentFunc()
+  {
+    var capture = Hud_GetChild( GetMenu( "ModListMenu" ), "MouseMovementCapture" )
+    AddMouseMovementCaptureHandler( capture, void function( int x, int y ) : ( capture ) { print( format( "registered mouse input from capture %s in x: %i; y: %i", capture.tostring(), x, y ) ) } )
+  }
+
```

Comment:
This section is a bit underwritten in my opinion. I know what you're talking about here because i happen to know what' you're doing, but it's not immediately obvious what "local relocalisation" is or why it's useful. Also, shouldn't this be called local relocation?

What i think you're trying to describe here is anonymous functions with implicit capture.

Reply:
I havr never heard the term 'implicit capture'? Nonetheless, nothing about this is implicit.
I can't think of anotger system in another language where you're required to explicitly transfer a local variable to a local anonymous function so my sleep deprived ass just made something up

Reply:
https://en.cppreference.com/w/cpp/language/lambda

See the section about Lambda Capture